### ISSUE TITLE
O(1) conversion to location of enum in enum_values 

### ISSUE THREAD

**Issue by ezyang** [Neargye/magic_enum#12]
Is there a way to do O(1) conversion of Color::BLUE to 1 (rather than 4, as `enum_integer` would do)?

Motivation: I have an enum generated by an external tool which may have gaps in it, and I need to pack the enums into a contiguous sequence of integers starting at 0 (so that I can more efficiently store it in memory somewhere else). I don't need this packing to be stable across invocations of the program, but it should be stable within the program.

**Comment by Neargye** [comment]
If I understand the question correctly, you can generate unordered_map to map enum -> index.

**Comment by Neargye** [comment]
```
constexpr auto enums= magic_enum::enum_values<E>();
std::unordered_map<E, std::size_t> map;
for (std::size_t i = 0; i < enums.size(); ++i ) {
 map.emplace(std::make_pair(enums[i], i));
}
```

**Comment by Neargye** [comment]
For getting map in compile-time can use https://github.com/serge-sans-paille/frozen.

**Comment by ezyang** [comment]
That's... a bit more kit than I would have expected, but I suppose it precisely answers the question :)

**Comment by Neargye** [comment]
Or you can use binaryserch, because enum values sorted.

**Comment by Neargye** [comment]
@ezyang I can add O(1) `enum_index(E value)` function if it's will be useful.
If enum value out of range, returns size enum or -1?

Example:
```cpp
enum class Color : int { RED = -10, BLUE = 0, GREEN = 10 };

 magic_enum::enum_index(Color::BLUE) -> 1?
 magic_enum::enum_index((Color)1) -> 3 or -1?
```

**Comment by ezyang** [comment]
@Neargye I am using your workarounds to implement in my project. If `enum_index` were added, I would use it instead, but in terms of absolute utility, I think this is more of an aesthetic question. In any case, out of bound errors should be consistent with your existing API, which I think implies that you should get an `optional<int>` out.

**Comment by Neargye** [comment]
Check dev branch, add `enum_index`, plus allows to get index at compile time.

**Comment by ezyang** [comment]
That is one sweet constexpr array! Thank you!

**Comment by Neargye** [comment]
Yes, it turned out a little easier than map :D

**Comment by alexkaratarakis** [comment]
Very nice addition, I am trying to do the same as @ezyang 
I am currently (at compile time) detecting if values are contiguous, and, if not, doing binary_search on `magic_enum::enum_values<T>()`, but this will make it even better and `O(1)`.

> Or you can use binaryserch, because enum values sorted.

Does this hold for all compilers?

**Comment by Neargye** [comment]
I add a test, and it's work on all compilers that support magic_enum(https://github.com/Neargye/magic_enum#compiler-compatibility).

I am currently working on stabilizing the new version, and I think this enum_index  will be included in the new release.

**Comment by Neargye** [comment]
Expected fix in v0.6.1

**Comment by Neargye** [comment]
Fix in master
